src/java/org/apache/cassandra/index/sai/utils/RangeIntersectionIterator.java

```diff
@@ -0,0 +1,229 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.cassandra.index.sai.utils;
+
+import java.io.IOException;
+import java.lang.invoke.MethodHandles;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.cassandra.io.util.FileUtils;
+import org.apache.cassandra.tracing.Tracing;
+
+/**
+ * A simple intersection iterator that makes no real attempts at optimising the iteration apart from
+ * initially sorting the ranges. This implementation also supports an intersection limit which limits
+ * the number of ranges that will be included in the intersection. This currently defaults to 2.
+ */
+@SuppressWarnings("resource")
+public class RangeIntersectionIterator extends RangeIterator
+{
+    private static final Logger logger = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+    // The cassandra.sai.intersection.clause.limit (default: 2) controls the maximum number of range iterator that
+    // will be used in the final intersection of a query operation.
+    private static final int INTERSECTION_CLAUSE_LIMIT = Integer.getInteger("cassandra.sai.intersection.clause.limit", 2);
+
+    static
+    {
+        logger.info(String.format("Storage attached index intersection clause limit is %d", INTERSECTION_CLAUSE_LIMIT));
+    }
+
+    private final List<RangeIterator> ranges;
+
+    private RangeIntersectionIterator(Builder.Statistics statistics, List<RangeIterator> ranges)
+    {
+        super(statistics);
+        this.ranges = ranges;
+    }
+
+    protected PrimaryKey computeNext()
+    {
+        // Range iterator that has been advanced in the previous cycle of the outer loop.
+        // Initially there hasn't been the previous cycle, so set to null.
+        int alreadyAvanced = -1;
+
+        // The highest primary key seen on any range iterator so far.
+        // It can become null when we reach the end of the iterator.
+        PrimaryKey highestKey = getCurrent();
+
+        outer:
+        while (highestKey != null && highestKey.compareTo(getMaximum()) <= 0)
```

Comment:
There's no way the `highestKey` should ever be greater than the maximum, right? Would this be better to either leave out or just make an assertion, which (if we used assertions properly) could be optionally disabled?

Reply:
Sorry about that. I added an assertion and then left the original code in place.